Source/Glide64/3dmath.cpp

```diff
@@ -151,48 +151,119 @@ void calc_sphere (VERTEX *v)
 float DotProductC(register float *v1, register float *v2)
 {
     register float result;
-    result = v1[0]*v2[0] + v1[1]*v2[1] + v1[2]*v2[2];
-    return(result);
+
+    result =
+        v1[0] * v2[0]
+      + v1[1] * v2[1]
+      + v1[2] * v2[2]
+    ;
+    return (result);
 }
 
 void NormalizeVectorC(float *v)
 {
     register float len;
-    len = sqrtf(v[0]*v[0] + v[1]*v[1] + v[2]*v[2]);
-    if (len > 0.0f)
-    {
-        v[0] /= len;
-        v[1] /= len;
-        v[2] /= len;
-    }
+
+    len = v[0]*v[0] + v[1]*v[1] + v[2]*v[2];
+    if (len == 0.0f)
+        return;
+
+    len = sqrtf(len); /* len >= 0, because a*a + b*b + c*c is never negative. */
+    v[0] /= len;
+    v[1] /= len;
+    v[2] /= len;
 }
 
 void TransformVectorC(float *src, float *dst, float mat[4][4])
 {
-  dst[0] = mat[0][0]*src[0] + mat[1][0]*src[1] + mat[2][0]*src[2];
-  dst[1] = mat[0][1]*src[0] + mat[1][1]*src[1] + mat[2][1]*src[2];
-  dst[2] = mat[0][2]*src[0] + mat[1][2]*src[1] + mat[2][2]*src[2];
+    float products[4][4];
+    register int i, j;
+    const int limit = 4 - 1; /* either xmm[127..0] or just xmm[95..0] works */
+
+#if 0
+    for (j = 0; j < limit; j++)
+        for (i = 0; i < limit; i++)
+            products[j][i] = mat[j][i] * src[j];
+#else
+    for (j = 0; j < limit; j++)
+        for (i = 0; i < limit; i++)
+            products[j][i]  = src[j]; /* _mm_set1_epi32(src[j]):  PSHUFDW */

```

Comment:
Because there's no such thing as a SSE2 instruction that multiplies rd = rs * rt all at once.  You need to divide the storage to its own step, then multiply by a source operand.  It's more organized--the divided split loops show these micro-steps better in the architecture's limitation.

And again, you're supposed to be able to change #if 0 to #if 1 in case you need to debug anything about the code during testing; that's the advantage over just /* commenting out */ all of your code.